server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -680,72 +689,156 @@ public String toString() {
-        ArrayUtil.timSort(files, Comparator.comparingLong(StoreFileMetaData::length)); // send smallest first
-        final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
-        final AtomicReference<Tuple<StoreFileMetaData, Exception>> error = new AtomicReference<>();
-        final byte[] buffer = new byte[chunkSizeInBytes];
-        for (final StoreFileMetaData md : files) {
-            if (error.get() != null) {
-                break;
+    void sendFiles(Store store, StoreFileMetaData[] files, Supplier<Integer> translogOps, ActionListener<Void> listener) {
+        final MultiFileSender multiFileSender = new MultiFileSender(store, files, translogOps);
+        resources.add(multiFileSender); // need to register to the resource list so we can clean up if the recovery gets cancelled.
+        final ActionListener<Void> wrappedListener = ActionListener.wrap(
+            r -> {
+                if (Assertions.ENABLED) {
+                    long expectedChunks = Arrays.stream(files).mapToLong(f -> (f.length() + chunkSizeInBytes - 1) / chunkSizeInBytes).sum();
+                    long sentChunks = multiFileSender.requestSeqIdTracker.getCheckpoint() + 1;
+                    assert sentChunks == expectedChunks : "sent chunks=" + sentChunks + " != expected chunks=" + expectedChunks;
+                }
+                multiFileSender.close();
+                listener.onResponse(null);
+            },
+            e -> {
+                IOUtils.closeWhileHandlingException(multiFileSender);
+                listener.onFailure(e);
+            });
+        multiFileSender.sendFileChunks(ActionListener.notifyOnce(wrappedListener));
+    }
+
+    private final class MultiFileSender implements Closeable {
+        private final Store store;
+        private final List<StoreFileMetaData> files;
+        private final Supplier<Integer> translogOps;
+        private final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
+        private final Deque<byte[]> recycledBuffers;
+        private boolean readerClosed; // ensure we don't reopen reader after the recovery was cancelled
+        private StoreFileMetaData md;
+        private InputStream currentInput;
+        private int position;
+        private final Object turnMutex = new Object();
+        private final AtomicBoolean turnAcquired = new AtomicBoolean();
+
+        MultiFileSender(Store store, StoreFileMetaData[] files, Supplier<Integer> translogOps) {
+            this.store = store;
+            this.files = new ArrayList<>(Arrays.asList(files));
+            this.translogOps = translogOps;
+            this.recycledBuffers = ConcurrentCollections.newDeque();
+            for (int i = 0; i < maxConcurrentFileChunks; i++) {
+                this.recycledBuffers.add(new byte[chunkSizeInBytes]);
```

Comment:
do we really have to pre-allocate these? I wonder if that's good practice, can we maybe not initialize it and instead allocate them lazily?

Reply:
okay